kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java

```diff
@@ -0,0 +1,218 @@
+ */
+public class KCBQThreadPoolExecutor extends ThreadPoolExecutor {
+
+  private static final Logger logger = LoggerFactory.getLogger(KCBQThreadPoolExecutor.class);
+
+  private ConcurrentHashMap<String, AtomicInteger> activeThreadTopicCount =
+      new ConcurrentHashMap<>();
+  private ConcurrentSet<Throwable> encounteredErrors = new ConcurrentSet<>();
+  private TopicPartitionManager topicPartitionManager;
+  private int topicThreadLimit;
+
+  /**
+   * @param config the {@link BigQuerySinkTaskConfig}
+   * @param context the {@link SinkTaskContext}
+   * @param workQueue the queue for storing tasks.
+   */
+  public KCBQThreadPoolExecutor(BigQuerySinkTaskConfig config,
+                                SinkTaskContext context,
+                                BlockingQueue<Runnable> workQueue) {
+    super(config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          // the following line is irrelevant because the core and max thread counts are the same.
+          1, TimeUnit.SECONDS,
+          workQueue);
+    topicPartitionManager = new TopicPartitionManager(context);
+    topicThreadLimit = config.getInt(BigQuerySinkTaskConfig.TOPIC_MAX_THREADS_CONFIG);
+  }
+
+  @Override
+  protected void beforeExecute(Thread thread, Runnable runnable) {
+    super.beforeExecute(thread, runnable);
+
+    if (runnable instanceof TableWriter) {
+      TableWriter tableWriter = (TableWriter) runnable;
+      String topic = tableWriter.getTopic();
+      activeThreadTopicCount.putIfAbsent(topic, new AtomicInteger(0));
+
+      int topicThreadCount = activeThreadTopicCount.get(topic).incrementAndGet();
+      if (topicThreadCount > topicThreadLimit) {
+        topicPartitionManager.pause(topic, topicThreadCount);
+      }
+    }
+  }
+
+  @Override
+  protected void afterExecute(Runnable runnable, Throwable throwable) {
+    super.afterExecute(runnable, throwable);
+
+    if (runnable instanceof TableWriter) {
+      TableWriter tableWriter = (TableWriter) runnable;
+      String topic = tableWriter.getTopic();
+
+      int topicThreadCount = activeThreadTopicCount.get(topic).decrementAndGet();
+      if (topicThreadCount < topicThreadLimit) {
+        topicPartitionManager.resume(topic);
+      }
+    }
+
+    if (throwable != null) {
+      logger.error("Task failed with {} error: {}",
+                   throwable.getClass().getName(),
+                   throwable.getMessage());
+      encounteredErrors.add(throwable);
+    }
+  }
+
+  /**
+   * Wait for all the currently queued tasks to complete, and then return.
+   *
+   * @throws BigQueryConnectException if any of the tasks failed.
+   * @throws InterruptedException if interrupted while waiting.
+   */
+  public void awaitCurrentTasks() throws InterruptedException, BigQueryConnectException {
+    int maximumPoolSize = getMaximumPoolSize();
+    CountDownLatch countDownLatch = new CountDownLatch(maximumPoolSize);
```

Comment:
Both join methods (by whatever means we would get the thread handles) and shutting down the executor rely on threads dying, which is not great because then we'll have to recreate them all when we end up putting again. (which is expensive)

The countdown latch makes sure that no threads need to die (and then be re-created expensively soon after).

if the concern is that the countdown is too confusing, I can add some comments around it.